File: client/src/data/record-batch.js
import { ProviderError, clone, inspectJson, uuid } from './data-provider.js';
import { MUTABLE_RECORD_FIELDS, patchRecord, recordReplacement } from './record-commands.js';
import { assertSourceWritable } from './record-schema.js';
import { normalizedTimes, validateRecord, validateRelationships, LOCAL_LIMITS } from './snapshot.js';

const mutable = new Set(MUTABLE_RECORD_FIELDS);
export const BATCH_BYTES = 8 * 1024 * 1024;

export function prepareRecordBatch(snapshot, operations, actor = 'local') {
  inspectJson({ operations });
  if (!Array.isArray(operations) || operations.length < 1 || operations.length > 500) throw new ProviderError('batch_capacity', 'Batch requires 1 to 500 operations', 413);
  if (new TextEncoder().encode(JSON.stringify({ operations })).length > BATCH_BYTES) throw new ProviderError('batch_capacity', 'Batch request exceeds 8 MiB', 413);
  const records = clone(snapshot.records), byId = new Map(records.map(record => [record.id, record])), seen = new Set(), items = [];
  for (const [index, operation] of operations.entries()) {
    try {
      if (!operation || typeof operation !== 'object' || Array.isArray(operation) || Object.keys(operation).some(key => !['type', 'recordId', 'expectedVersion', 'payload'].includes(key))) throw new ProviderError('invalid_batch', 'Unsupported batch operation fields');
      const type = operation.type;
      if (!['create', 'update', 'replace', 'patch', 'delete', 'restore'].includes(type)) throw new ProviderError('invalid_command', 'Unsupported batch operation');
      let record, payload = clone(operation.payload === undefined ? {} : operation.payload);
      const now = new Date().toISOString();
      if (type === 'create') {
        if (Object.hasOwn(operation, 'recordId') || Object.hasOwn(operation, 'expectedVersion')) throw new ProviderError('immutable_field', 'Create IDs and versions are provider-owned');
        record = { id: uuid(), workspaceId: snapshot.manifest.workspaceId, kind: 'event', title: '', start: now, end: null,
          parentSessionId: null, order: 0, sourceId: snapshot.manifest.scope.sourceIds[0], groupIds: [], tags: [], data: {},
          render: { color: '#39788a' }, extensions: {}, schemaId: null, schemaVersion: null, originalStart: null, originalEnd: null,
          version: 1, createdAt: now, updatedAt: now, createdBy: actor, updatedBy: actor, deletedAt: null };
      } else {
        if (typeof operation.recordId !== 'string' || !/^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/u.test(operation.recordId)) throw new ProviderError('invalid_id', 'Record IDs must be canonical UUIDs');
        if (seen.has(operation.recordId)) throw new ProviderError('duplicate_batch_record', 'A record may occur only once in a batch');
        seen.add(operation.recordId); record = byId.get(operation.recordId);
        if (!record) throw new ProviderError('record_not_found', 'Record does not exist', 404);
        if (!Number.isSafeInteger(operation.expectedVersion)) throw new ProviderError('precondition_required', 'Every existing record requires expectedVersion', 428);
        if (operation.expectedVersion !== record.version) throw new ProviderError('version_conflict', 'A record changed; refresh before retrying', 412);
        if (record.version >= Number.MAX_SAFE_INTEGER) throw new ProviderError('revision_capacity', 'Record version capacity reached', 413);
        if (record.deletedAt && type !== 'restore') throw new ProviderError('record_deleted', 'Record is deleted', 409);
        assertSourceWritable(snapshot, record.sourceId, type);
        if (type === 'patch') payload = patchRecord(record, payload);
        if (type === 'replace') payload = recordReplacement(payload);
      }
      if (!payload || typeof payload !== 'object' || Array.isArray(payload)) throw new ProviderError('invalid_record', 'Record payload must be an object');
      if (['delete', 'restore'].includes(type)) {
        if (Object.keys(payload).length) throw new ProviderError('invalid_request', 'Delete and restore do not accept record fields');
        if (type === 'restore' && !record.deletedAt) throw new ProviderError('record_not_deleted', 'Only deleted records can be restored', 409);
        record.deletedAt = type === 'delete' ? now : null;
      } else {
        if (Object.keys(payload).some(key => !mutable.has(key))) throw new ProviderError('immutable_field', 'Record identity, versions and audit fields are provider-owned');
        const sourceId = Object.hasOwn(payload, 'sourceId') ? payload.sourceId : record.sourceId;
        const policy = assertSourceWritable(snapshot, sourceId, type === 'create' ? 'create' : sourceId !== record.sourceId ? 'reassign' : 'update');
        if (type === 'create' && !Object.hasOwn(payload, 'schemaId') && !Object.hasOwn(payload, 'schemaVersion') && policy.defaultSchema) Object.assign(payload, { schemaId: policy.defaultSchema.id, schemaVersion: policy.defaultSchema.version });
        if (type !== 'create' && Object.hasOwn(payload, 'kind') && payload.kind !== record.kind) throw new ProviderError('immutable_kind', 'Record kind cannot be changed');
        Object.assign(record, payload);
      }
      if (type !== 'create') { record.version++; record.updatedAt = now; record.updatedBy = actor; }
      normalizedTimes(record); validateRecord(record, snapshot);
      if (record.groupIds.some(id => !snapshot.groups.some(group => group.id === id))) throw new ProviderError('group_unavailable', 'Record group is not in the workspace catalog');
      if (type === 'create') { records.push(record); byId.set(record.id, record); }
      items.push({ index, record });
    } catch (error) {
      if (error instanceof ProviderError) error.errors = [{ index, code: error.code, message: error.message }];
      throw error;
    }
  }
  if (records.length > LOCAL_LIMITS.records) throw new ProviderError('record_limit', 'Local record capacity reached', 413);
  try { validateRelationships(records, snapshot.manifest.workspaceId, snapshot.manifest.scope.sourceIds, snapshot); }
  catch (error) {
    if (error instanceof ProviderError) error.errors = [{ index: null, code: error.code, message: 'The final batch would leave an invalid parent relationship' }];
    throw error;
  }
  return { records, items };
}

File: server/app/models/record_batch.py
"""Prepare a complete candidate without performing I/O or publishing state."""
from __future__ import annotations

import copy

from .domain import DomainError, MAX_SAFE_INT, json_bytes, make_record, now_iso, valid_id, validate_json, validate_record
from .record_commands import patch_record, replacement
from .record_schema import assert_source_writable

BATCH_BYTES = 8 * 1024 * 1024
BATCH_RECORDS = 500


def batch_operations(payload):
    validate_json(payload)
    if len(json_bytes(payload)) > BATCH_BYTES:
        raise DomainError("batch_capacity", "Batch request exceeds 8 MiB.", 413)
    if not isinstance(payload, dict) or set(payload) != {"operations"}:
        raise DomainError("invalid_batch", "Batch body requires only operations.")
    operations = payload["operations"]
    if not isinstance(operations, list) or not 1 <= len(operations) <= BATCH_RECORDS:
        raise DomainError("batch_capacity", "Batch requires 1 to 500 operations.", 413)
    return operations


def prepare_batch(records, configuration, payload, actor, authorize=None):
    operations = batch_operations(payload)
    workspace = configuration["manifest"]["workspaceId"]
    candidate, changed, seen = dict(records), [], set()
    for index, operation in enumerate(operations):
        try:
            if not isinstance(operation, dict) or set(operation) - {"type", "recordId", "expectedVersion", "payload"}:
                raise DomainError("invalid_batch", "Unsupported batch operation fields.")
            kind = operation.get("type")
            if kind not in ("create", "update", "replace", "patch", "delete", "restore"):
                raise DomainError("invalid_command", "Unsupported batch operation.")
            previous = None
            if kind == "create":
                if "recordId" in operation or "expectedVersion" in operation:
                    raise DomainError("immutable_field", "Create IDs and versions are provider-owned.")
            else:
                record_id = valid_id(operation.get("recordId"))
                if record_id in seen:
                    raise DomainError("duplicate_batch_record", "A record may occur only once in a batch.")
                seen.add(record_id)
                previous = records.get(record_id)
                if previous is None:
                    raise DomainError("record_not_found", "Record does not exist.", 404)
                if authorize:
                    authorize(kind, previous)
                expected = operation.get("expectedVersion")
                if isinstance(expected, bool) or not isinstance(expected, int):
                    raise DomainError("precondition_required", "Every existing record requires expectedVersion.", 428)
                if previous["version"] != expected:
                    raise DomainError("version_conflict", "A record changed; refresh before retrying.", 412)
                if previous["version"] >= MAX_SAFE_INT:
                    raise DomainError("revision_capacity", "Record version capacity reached.", 413)
                if previous["deletedAt"] and kind != "restore":
                    raise DomainError("record_deleted", "Record is deleted.", 409)
                assert_source_writable(configuration, previous["sourceId"], kind)
            assigned = copy.deepcopy(operation.get("payload", {}))
            if kind == "patch":
                assigned = patch_record(previous, assigned)
            elif kind == "replace":
                assigned = replacement(assigned)
            if not isinstance(assigned, dict):
                raise DomainError("invalid_record", "Record payload must be an object.")
            if kind in ("delete", "restore"):
                if assigned:
                    raise DomainError("invalid_request", "Delete and restore do not accept record fields.")
                record = copy.deepcopy(previous)
                if kind == "restore" and not previous["deletedAt"]:
                    raise DomainError("record_not_deleted", "Only deleted records can be restored.", 409)
                timestamp = now_iso()
                record.update(version=record["version"] + 1, updatedAt=timestamp, updatedBy=actor,
                              deletedAt=timestamp if kind == "delete" else None)
            else:
                if kind == "create" and "sourceId" not in assigned:
                    sources = configuration["manifest"]["scope"]["sourceIds"]
                    if not sources:
                        raise DomainError("source_unavailable", "Configure a source before creating records.")
                    assigned["sourceId"] = sources[0]
                source_id = assigned.get("sourceId", previous["sourceId"] if previous else None)
                policy = assert_source_writable(configuration, source_id, "create" if previous is None else
                                                "reassign" if source_id != previous["sourceId"] else "update")
                if kind == "create" and "schemaId" not in assigned and "schemaVersion" not in assigned and policy.get("defaultSchema"):
                    assigned.update(schemaId=policy["defaultSchema"]["id"], schemaVersion=policy["defaultSchema"]["version"])
                record = make_record(assigned, actor, workspace, previous)
            if authorize:
                authorize(kind, record)
            validate_record(record, workspace=workspace, configuration=configuration)
            candidate[record["id"]] = record
            changed.append({"index": index, "record": record})
        except DomainError as error:
            error.errors = [{"index": index, "code": error.code, "message": error.message}]
            raise
    changed_ids = {item["record"]["id"]: item["index"] for item in changed}
    # Parent edits can invalidate unchanged descendants, so validate the final graph.
    from .domain import _validate_record_relationships
    for record in candidate.values():
        try:
            _validate_record_relationships(record, candidate, workspace)
        except DomainError as error:
            related = record["id"] if record["id"] in changed_ids else record.get("parentSessionId")
            error.errors = [{"index": changed_ids.get(related), "code": error.code,
                             "message": "The final batch would leave an invalid parent relationship."}]
            raise
    return candidate, changed
